awx/main/dispatch/__init__.py

```diff
@@ -68,10 +77,23 @@ def pg_bus_conn(new_connection=False):
         conn.set_session(autocommit=True)
     else:
         if pg_connection.connection is None:
-            pg_connection.connect()
-        if pg_connection.connection is None:
-            raise RuntimeError('Unexpectedly could not connect to postgres for pg_notify actions')
-        conn = pg_connection.connection
+            for retry_count in range(MAX_RETRIES):
+                try:
+                    pg_connection.connect()
+                    conn = pg_connection.connection
+                except (psycopg2.OperationalError, OperationalError, Error) as reconn_error:
+                    logger.warning(f'Database unavailable. Retry with new connection in next {POLL_SECONDS} seconds. Attempt {retry_count}/{MAX_RETRIES}.')
+                    time.sleep(POLL_SECONDS)
+                    retry_conn = True
+                else: 
+                    if retry_conn:
+                        logger.warning('Run dispatcher restart due to database connection loss and restore.')
+                        os.kill(os.getppid(), signal.SIGTERM)
```

Comment:
If you need to break out of the dispatcher main loop here

https://github.com/ansible/awx/blob/0815f935ca359537c023efadf286671e31dc3374/awx/main/dispatch/worker/base.py#L175

You could do that by throwing _any_ exception that's not one of them that gets caught. You could throw a `RuntimeError` or a pure `Exception` and this will exit the dispatcher, and then the decision-making about what to do after that (like restarting) is punted to supervisor then systemctl / kubernetes.

Reply:
I notice that error arrive on pg_connection.connect() and  i need to check if with raise flow arrive to base.py. Tomorrow I will try this option. 

What i can say now is that i tried to avoid the "kill" of PID with reause pg_connection, but seems that a down of postgres broke something on what was already in place in terms of connection and queue listener. 

Any suggestion?

Reply:
changed approach